src/future/batch_semaphore.rs

```diff
@@ -0,0 +1,604 @@
+                    num_permits -= *batch_size;
+                    permit_clocks.pop_front();
+                }
+
+                // Break early to avoid causally depending on the next batch.
+                if num_permits == 0 {
+                    break;
+                }
+            }
+
+            assert_eq!(num_permits, 0);
+            Ok(clock)
+        } else {
+            // There are not enough permits to fulfill the request.
+            Err(TryAcquireError::NoPermits)
+        }
+    }
+
+    fn release(&mut self, num_permits: usize, clock: VectorClock) {
+        self.init_permit_clocks();
+        self.num_available += num_permits;
+        self.permit_clocks.as_mut().unwrap().push_back((num_permits, clock));
+    }
+}
+
+/// Fairness mode for the semaphore. Determines which threads are woken when
+/// permits are released.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Fairness {
+    /// The semaphore is strictly fair, so earlier requesters always get
+    /// priority over later ones.
+    StrictlyFair,
+
+    /// The semaphore makes no guarantees about fairness. In particular,
+    /// a waiter can be starved by other threads.
+    Unfair,
+}
+
+/// A counting semaphore which permits waiting on multiple permits at once,
+/// and supports both asychronous and synchronous blocking operations.
+#[derive(Debug)]
+struct BatchSemaphoreState {
+    // Key invariants:
+    //
+    // (1) if `waiters` is nonempty and the head waiter is `H`,
+    // then `H.num_permits > permits_available.available()`.  (In other words,
+    // we are never in a state where there are enough permits available for the
+    // first waiter.  This invariant is ensured by the `drop` handler below.)
+    //
+    // (2) W is in waiters iff W.is_queued
+    //
+    // (3) W.is_queued ==> !W.has_permits
+    // Note: the converse is not true.  We can have !W.has_permits && !W.is_queued
+    // when the Acquire is created but not yet polled.
+    //
+    // (4) closed ==> waiters.is_empty()
+    waiters: VecDeque<Arc<Waiter>>,
+    permits_available: PermitsAvailable,
+    fairness: Fairness,
+    // TODO: should there be a clock for the close event?
+    closed: bool,
+}
+
+impl BatchSemaphoreState {
+    fn acquire_permits(&mut self, num_permits: usize) -> Result<(), TryAcquireError> {
+        assert!(num_permits > 0);
+
+        // Permits here can be acquired in one of two scenarios:
+        // - The waiter queue is empty; nobody else is waiting for permits,
+        //   so if there are enough available, immediately succeed.
+        // - The semaphore is operating in an unfair mode; the current thread
+        //   is either requesting permits for the first time, or it was woken
+        //   and selected by the scheduler. In either case, the thread may
+        //   succeed, as long as there are enough permits.
+        let can_acquire = self.waiters.is_empty() || matches!(self.fairness, Fairness::Unfair);
```

Comment:
I find this style of binding when there's no need odd. Is it only to have something to attach the comment to or am I missing something?

Reply:
Why isn't the `if self.closed` branch right under the `assert!`?

Reply:
Yes, I just want it to be clear what the comment is about. Inlining this into the condition might make it unclear: is the comment about the overall code in that condition? I think in this case (and the `try_to_acquire` case) it is easier to read.

Reply:
I may have agreed with your reasoning if the branch actually contained any code, and here it becomes clear from the `Permits here can be acquired in one of two scenarios:` line.

Regardless: I think the `if self.closed` branch should be moved up. The current placement adds cognitive overhead when reading the function top to bottom.